### PR TITLE
MAILET-115 Refactor AbstractRedirect & AbstractNotify based mailets

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

**Comment by aduprat** [comment]
test this please

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/NotifySender.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.transport.mailets;
+package org.apache.james.transport.mailets.redirect;
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/NotifySender.java]: notify is not a redirect, isn't it ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/InitParameters.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets.redirect;
+
+public interface InitParameters {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/InitParameters.java]: all redirect mails use all these fields ?

File: server/mailet/mailets/pom.xml [hunk 3]
```diff
@@ -181,6 +181,12 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-guava</artifactId>
+            <version>1.3.1</version>
```
Comment (mbaechler) [inline server/mailet/mailets/pom.xml]: don't we already have a version for that library into server/pom.xml ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    private final MailetContext mailetContext;
+
+    public AddressExtractor(MailetContext mailetContext) {
+        this.mailetContext = mailetContext;
+    }
+
+    public List<MailAddress> from(String addressList, List<String> allowedSpecials) throws MessagingException {
+        try {
+            return toMailAddresses(ImmutableList.copyOf(InternetAddress.parse(addressList, false)), allowedSpecials);
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java]: you don't need to copy the list, you are the only one the have a reference on it. Or does it generate a raw array ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+public class AddressExtractor {
+
+    private final MailetContext mailetContext;
+
+    public AddressExtractor(MailetContext mailetContext) {
+        this.mailetContext = mailetContext;
+    }
+
+    public List<MailAddress> from(String addressList, List<String> allowedSpecials) throws MessagingException {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java]: this API is a bit ugly

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/NotifySender.java (hunk 1 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/NotifySender.java]: Moved

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/InitParameters.java (hunk 2 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/InitParameters.java]: No

File: server/mailet/mailets/pom.xml (hunk 3 above)
Comment (aduprat) [inline server/mailet/mailets/pom.xml]: OK

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java (hunk 4 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java]: Bad java mail API which return a raw array...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java (hunk 5 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java]: Can you provide a better one?

Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AddressExtractor.java]: `AddressExtractor.withContext(context).allowedSpecials(blabla).extract(addressList)`

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/utils/MimeMessageModifier.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.apache.james.transport.mailets.utils;
+
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Joiner;
+
+public class MimeMessageModifier {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/utils/MimeMessageModifier.java]: why won't you pass the message in the constructor ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/StaticInitParameters.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets.redirect;
+
+public class StaticInitParameters implements InitParameters {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/StaticInitParameters.java]: what do you mean by "static" ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AbstractRedirect.java [hunk 8]
```diff
@@ -817,34 +563,30 @@ protected void setIsReply(Mail newMail, boolean isReply, Mail originalMail) thro
...
-            subject = getSubject();
-            subjectPrefix = getSubjectPrefix();
-            apparentlyTo = getTo();
-            attachError = attachError();
-            isReply = isReply();
-            if (isDebug) {
-                String logBuffer = "static" + ", passThrough=" + passThrough + ", fakeDomainCheck=" + fakeDomainCheck + ", sender=" + sender + ", replyTo=" + replyTo + ", reversePath=" + reversePath + ", message=" + messageText + ", recipients=" + arrayToString(recipients == null ? null : recipients.toArray()) + ", subject=" + subject + ", subjectPrefix=" + subjectPrefix + ", apparentlyTo=" + arrayToString(apparentlyTo) + ", attachError=" + attachError + ", isReply=" + isReply + ", attachmentType=" + attachmentType + ", inLineType=" + inLineType + " ";
+        if (getInitParameters().isStatic()) {
+            if (getInitParameters().isDebug()) {
+                String logBuffer = "static" + ", passThrough=" + getInitParameters().getPassThrough() +
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AbstractRedirect.java]: why it's not the responsibility of your InitParameters object ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/StaticInitParameters.java (hunk 7 above)
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/StaticInitParameters.java]: It's not obvious what the purpose is. Can you clarify that ?

Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/StaticInitParameters.java]: Done

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/CharsetFromSubjectMailHeader.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+     * cannot determine or the the text is not encoded then it returns null.
+     * <p/>
+     * Here is an example raw text: Subject:
+     * =?iso-8859-2?Q?leg=FAjabb_pr=F3ba_l=F5elemmel?=
+     *
+     * @param subject the raw (not decoded) value of the header. Null means that the
+     *                header was not present (in this case it always return null).
+     * @return the MIME charset name or null if no encoding applied
+     */
+    public Optional<String> parse(String subject) {
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/CharsetFromSubjectMailHeader.java]: There is nothing is mime4j to do that?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/SpecialAddressTest.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Test
+    public void nullSpecialAddressShouldMatchExpectedValue() {
+        assertThat(SpecialAddress.AddressMarker.NULL).isEqualTo("null@address.marker");
+    }
+
+    @Test
+    public void isSpecialAddressShouldReturnTrueWhenMatchingSpecialDomain() throws Exception {
+        assertThat(SpecialAddress.isSpecialAddress(new MailAddress("user", "address.marker")));
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/SpecialAddressTest.java]: assertThat what? :)

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/SpecialAddressTest.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Test
+    public void isSpecialAddressShouldReturnFalseWhenNotMatchingSpecialDomain() throws Exception {
+        assertThat(SpecialAddress.isSpecialAddress(new MailAddress("user", "james.org")));
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/SpecialAddressTest.java]: idem

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Bounce.java [hunk 12]
```diff
@@ -44,7 +51,7 @@
  * A notice text can be specified, and in such case will be inserted into the
  * notification inline text.<br>
  * If the notified message has an "error message" set, it will be inserted into
- * the notification inline text. If the <code>attachStackTrace</code> init
+ * the notification inline text. If the <code>attachError</code> init
```
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Bounce.java]: We can, if we change the documentation.
Having an unknown parameter will produce an error when starting James with the parameter name.
So users can easily find which is the mailet, and follow the documentation.

This parameter (`attachStackTrace`) was an alias of (`attachError`) used in other mailets.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/CharsetFromSubjectMailHeader.java (hunk 9 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/CharsetFromSubjectMailHeader.java]: No, it's a very particular treatment, that's why a keep the documentation.

**Comment by aduprat** [comment]
Replaced by #423

**Pull request closed without merging**